src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -59,7 +66,63 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
         {
             lock (_gate)
             {
-                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+                var documentId = e.Document.Id;
+                var filePath = e.Document.FilePath;
+                _openDocumentContexts.Add(documentId, Task.Run(async () =>
+                {
+                    var context = await GetConventionContextAsync(filePath, CancellationToken.None).ConfigureAwait(false);
+                    context.CodingConventionsChangedAsync += (innerSender, innerE) => HandleCodingConventionsChangedAsync(documentId, innerSender, innerE);
+                    return context;
+                }));
+            }
+        }
+
+        private static readonly ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>> s_projectNotifications =
+            new ConditionalWeakTable<CodingConventionsChangedEventArgs, HashSet<ProjectId>>();
+
+        private Task HandleCodingConventionsChangedAsync(DocumentId documentId, object sender, CodingConventionsChangedEventArgs e)
+        {
+            var projectId = documentId.ProjectId;
+            var projectsAlreadyNotified = s_projectNotifications.GetOrCreateValue(e);
+            lock (projectsAlreadyNotified)
+            {
+                if (!projectsAlreadyNotified.Add(projectId))
+                {
+                    return Task.CompletedTask;
+                }
+            }
+
+#if EDITORFEATURESWPF
+            var foregroundNotificationService = ((IMefHostExportProvider)_workspace.Services.HostServices).GetExports<IForegroundNotificationService>().Single().Value;
+            foregroundNotificationService.RegisterNotification(UpdateProject, EmptyAsyncToken.Instance, CancellationToken.None);
+#else
+#endif
+
+            return Task.CompletedTask;
+
+            void UpdateProject()
+            {
+                var compilationOptions = _workspace.CurrentSolution.GetProject(projectId)?.CompilationOptions;
```

Comment:
you have document id, why refresh whole project rather than just document?

Reply:
➡️ The current implementation uses this for two reasons:

1. We only get events for open files, but most .editorconfig options apply to at least a project. This gets us *close* to a full working experience for users with FSA on.
2. This was the finest granularity event I saw that appears to have the intended impact.